Elasticsearch 8.13 Percolator Regression, Exception thrown On Wildcard Query createWeight

### Elasticsearch Version

8.13.1

### Installed Plugins

Percolator

### Java Version

17

### OS Version

N/A

### Problem Description

Upon upgrading from Elasticsearch 8.12 to 8.13.1, the code shown in the "Steps to Reproduce" section below started throwing an exception when it used to pass (using the Java client API). I’m running a single node Elasticsearch cluster. Is this a regression, or is there an expected API break in the way Percolator is meant to be used on ES 8.13?

### Steps to Reproduce

Run this code with an Elasticsearch client pointing at any cluster:

```
    private static void basicPercolatorFailureExample(co.elastic.clients.elasticsearch.ElasticsearchClient client) throws IOException {
        client
                .indices()
                .create(builder -> builder
                        .index("test-docs-index")
                        .mappings(mappingBuilder -> mappingBuilder
                                .properties("field1", propBuilder -> propBuilder.keyword(kwBuilder -> kwBuilder))));
        client.indices()
                .create(builder -> builder
                        .index("test-percolate-index")
                        .mappings(mappingBuilder -> mappingBuilder
                                .properties("field1", propBuilder -> propBuilder.keyword(kwBuilder -> kwBuilder))
                                .properties(
                                        "query",
                                        propBuilder -> propBuilder.percolator(percBuilder -> percBuilder))));
        client.index(builder -> builder
                .id("id")
                .document(Map.of("field1", "doc_1"))
                .index("test-docs-index")
                .refresh(Refresh.True));
        Query wildcardQuery = Query.of(builder -> builder.wildcard(
                wcBuilder -> wcBuilder.field("field1").queryName("field1").value("doc_*")));
        IndexRequest<?> indexRequest = IndexRequest.of(builder -> builder
                .id("percolator-id")
                .index("test-percolate-index")
                .document(new PercolatorDocument(wildcardQuery))
                .refresh(Refresh.True));
        client.index(indexRequest);
        // Thw following code throws an exception
        SearchResponse<?> percolateResponse = client.search(
                sb -> sb
                        .query(qb -> qb
                                .percolate(perc -> perc
                                        .index("test-docs-index")
                                        .id("id")
                                        .field("query")))
                        .index("test-percolate-index"),
                JsonData.class);

        // This code is never reached
        org.assertj.core.api.Assertions.assertThat(percolateResponse.hits().total().value()).isOne();
    }
```

Exception thrown on the client:

```
org.elasticsearch.client.ResponseException: method [POST], host [https://<REDACTED>:9200], URI [/test-percolate-index/_search?typed_keys=true], status line [HTTP/1.1 500 Internal Server Error]
{"error":{"root_cause":[{"type":"unsupported_operation_exception","reason":"Query field1:doc_* does not implement createWeight"}],"type":"search_phase_execution_exception","reason":"all shards failed","phase":"query","grouped":true,"failed_shards":[{"shard":0,"index":"test-percolate-index","node":"jF9bdyvkSuK716u0rUVk4w","reason":{"type":"unsupported_operation_exception","reason":"Query field1:doc_* does not implement createWeight"}}],"caused_by":{"type":"unsupported_operation_exception","reason":"Query field1:doc_* does not implement createWeight","caused_by":{"type":"unsupported_operation_exception","reason":"Query field1:doc_* does not implement createWeight"}}},"status":500}
    at org.elasticsearch.client.RestClient.convertResponse(RestClient.java:351)
    at org.elasticsearch.client.RestClient.performRequest(RestClient.java:317)
    at org.elasticsearch.client.RestClient.performRequest(RestClient.java:292)
    at co.elastic.clients.transport.rest_client.RestClientHttpClient.performRequest(RestClientHttpClient.java:91)
    at co.elastic.clients.transport.ElasticsearchTransportBase.performRequest(ElasticsearchTransportBase.java:144)
    at co.elastic.clients.elasticsearch.ElasticsearchClient.search(ElasticsearchClient.java:1923)
    at co.elastic.clients.elasticsearch.ElasticsearchClient.search(ElasticsearchClient.java:1940)
```

Exception thrown on the server:

```
DEBUG [2024-04-02T17:51:52.605894Z] org.elasticsearch.action.search.TransportSearchAction: [jF9bdyvkSuK716u0rUVk4w][test-percolate-index][0]: Failed to execute [SearchRequest{searchType=QUERY_THEN_FETCH, indices=[test-percolate-index], indicesOptions=IndicesOptions[ignore_unavailable=false, allow_no_indices=true, expand_wildcards_open=true, expand_wildcards_closed=false, expand_wildcards_hidden=false, allow_aliases_to_multiple_indices=true, forbid_closed_indices=true, ignore_aliases=false, ignore_throttled=true], routing='null', preference='null', requestCache=null, scroll=null, maxConcurrentShardRequests=0, batchedReduceSize=512, preFilterShardSize=null, allowPartialSearchResults=true, localClusterAlias=null, getOrCreateAbsoluteStartMillis=-1, ccsMinimizeRoundtrips=true, source={"query":{"percolate":{"field":"query","documents":[{"field1":"doc_1"}]}}}}] lastShard [true] (0: [jF9bdyvkSuK716u0rUVk4w][test-percolate-index][0]: Failed to execute [SearchRequest{searchType=QUERY_THEN_FETCH, indices=[test-percolate-index], indicesOptions=IndicesOptions[ignore_unavailable=false, allow_no_indices=true, expand_wildcards_open=true, expand_wildcards_closed=false, expand_wildcards_hidden=false, allow_aliases_to_multiple_indices=true, forbid_closed_indices=true, ignore_aliases=false, ignore_throttled=true], routing='null', preference='null', requestCache=null, scroll=null, maxConcurrentShardRequests=0, batchedReduceSize=512, preFilterShardSize=null, allowPartialSearchResults=true, localClusterAlias=null, getOrCreateAbsoluteStartMillis=-1, ccsMinimizeRoundtrips=true, source={"query":{"percolate":{"field":"query","documents":[{"field1":"doc_1"}]}}}}] lastShard [true], throwable0_message: [<REDACTED>][172.29.0.3:9300][indices:data/read/search[phase/query]], throwable1_message: Fetch Failed [Error running fetch phase for doc [0]], throwable2_message: Query field1:doc_* does not implement createWeight)
org.elasticsearch.transport.RemoteTransportException: [<REDACTED>][172.29.0.3:9300][indices:data/read/search[phase/query]]
Caused by: org.elasticsearch.search.fetch.FetchPhaseExecutionException: Fetch Failed [Error running fetch phase for doc [0]]
        at org.elasticsearch.server@8.13.0/org.elasticsearch.search.fetch.FetchPhaseDocsIterator.iterate(FetchPhaseDocsIterator.java:78)
        at org.elasticsearch.server@8.13.0/org.elasticsearch.search.fetch.FetchPhase.buildSearchHits(FetchPhase.java:170)
        at org.elasticsearch.server@8.13.0/org.elasticsearch.search.fetch.FetchPhase.execute(FetchPhase.java:78)
        at org.elasticsearch.server@8.13.0/org.elasticsearch.search.SearchService.executeFetchPhase(SearchService.java:720)
        at org.elasticsearch.server@8.13.0/org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:691)
        at org.elasticsearch.server@8.13.0/org.elasticsearch.search.SearchService.lambda$executeQueryPhase$2(SearchService.java:549)
        at org.elasticsearch.server@8.13.0/org.elasticsearch.action.ActionRunnable$3.accept(ActionRunnable.java:73)
        at org.elasticsearch.server@8.13.0/org.elasticsearch.action.ActionRunnable$3.accept(ActionRunnable.java:70)
        at org.elasticsearch.server@8.13.0/org.elasticsearch.action.ActionRunnable$4.doRun(ActionRunnable.java:95)
        at org.elasticsearch.server@8.13.0/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
        at org.elasticsearch.server@8.13.0/org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:33)
        at org.elasticsearch.server@8.13.0/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:984)
        at org.elasticsearch.server@8.13.0/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: java.lang.UnsupportedOperationException: Query field1:doc_* does not implement createWeight
        at org.apache.lucene.core@9.10.0/org.apache.lucene.search.Query.createWeight(Query.java:84)
        at org.apache.lucene.core@9.10.0/org.apache.lucene.search.NamedMatches$NamedQuery.createWeight(NamedMatches.java:101)
        at org.apache.lucene.core@9.10.0/org.apache.lucene.search.IndexSearcher.createWeight(IndexSearcher.java:874)
        at org.elasticsearch.percolator@8.13.0/org.elasticsearch.percolator.PercolatorMatchedSlotSubFetchPhase$1.process(PercolatorMatchedSlotSubFetchPhase.java:108)
        at org.elasticsearch.server@8.13.0/org.elasticsearch.search.fetch.FetchPhase$1.nextDoc(FetchPhase.java:164)
        at org.elasticsearch.server@8.13.0/org.elasticsearch.search.fetch.FetchPhaseDocsIterator.iterate(FetchPhaseDocsIterator.java:70)
        ... 15 more
```


### Logs (if relevant)

_No response_

--- a/docs/changelog/107432.yaml
+++ b/docs/changelog/107432.yaml
@@ -1,0 +1,6 @@
+pr: 107432
+summary: "Percolator named queries: rewrite for matched info"
+area: Percolator
+type: bug
+issues:
+ - 107176

--- a/modules/percolator/src/main/java/org/elasticsearch/percolator/PercolatorMatchedSlotSubFetchPhase.java
+++ b/modules/percolator/src/main/java/org/elasticsearch/percolator/PercolatorMatchedSlotSubFetchPhase.java
@@ -85,8 +85,9 @@ final class PercolatorMatchedSlotSubFetchPhase implements FetchSubPhase {
                         // This is not a document with a percolator field.
                         continue;
                     }
-                    query = pc.filterNestedDocs(query, fetchContext.getSearchExecutionContext().indexVersionCreated());
                     IndexSearcher percolatorIndexSearcher = pc.percolateQuery.getPercolatorIndexSearcher();
+                    query = pc.filterNestedDocs(query, fetchContext.getSearchExecutionContext().indexVersionCreated());
+                    query = percolatorIndexSearcher.rewrite(query);
                     int memoryIndexMaxDoc = percolatorIndexSearcher.getIndexReader().maxDoc();
                     TopDocs topDocs = percolatorIndexSearcher.search(query, memoryIndexMaxDoc, new Sort(SortField.FIELD_DOC));
                     if (topDocs.totalHits.value == 0) {

--- a/modules/percolator/src/yamlRestTest/resources/rest-api-spec/test/30_matched_complex_queries.yml
+++ b/modules/percolator/src/yamlRestTest/resources/rest-api-spec/test/30_matched_complex_queries.yml
@@ -1,0 +1,86 @@
+setup:
+  - requires:
+      cluster_features: ["gte_v8.14.0"]
+      reason: "Displaying matched complex named queries within percolator queries was fixed in 8.14"
+  - do:
+      indices.create:
+        index: houses
+        body:
+          mappings:
+            dynamic: strict
+            properties:
+              my_query:
+                type: percolator
+              description:
+                type: text
+              num_of_bedrooms:
+                type: integer
+              type:
+                type: keyword
+              price:
+                type: integer
+
+  - do:
+      index:
+        refresh: true
+        index: houses
+        id: query_cheap_houses_with_swimming_pool
+        body:
+          my_query:
+            {
+              "bool": {
+                "should": [
+                  { "range": { "price": { "lte": 399999, "_name": "cheap_query" } } },
+                  { "wildcard": { "description": { "value": "swim*", "_name": "swimming_pool_query" } } }
+                ]
+              }
+            }
+
+  - do:
+      index:
+        refresh: true
+        index: houses
+        id: query_big_houses_with_fireplace
+        body:
+          my_query:
+            {
+              "bool": {
+                "should": [
+                  { "range": { "num_of_bedrooms": { "gte": 3, "_name": "big_house_query" } } },
+                  { "query_string": { "query": "fire*", "fields" : ["description"],  "_name": "fireplace_query" } }
+                ]
+              }
+            }
+
+---
+"Matched named queries within percolator queries: percolate existing document":
+  - do:
+      index:
+        refresh: true
+        index: houses
+        id: house1
+        body:
+          description: "house with a beautiful fireplace and swimming pool"
+          num_of_bedrooms: 3
+          type: detached
+          price: 1000000
+
+  - do:
+      search:
+        index: houses
+        body:
+          query:
+            percolate:
+              field: my_query
+              index: houses
+              id: house1
+
+  - match: { hits.total.value: 2 }
+
+  - match: { hits.hits.0._id: query_big_houses_with_fireplace }
+  - match: { hits.hits.0.fields._percolator_document_slot: [ 0 ] }
+  - match: { hits.hits.0.fields._percolator_document_slot_0_matched_queries: [ "big_house_query", "fireplace_query" ] }
+
+  - match: { hits.hits.1._id: query_cheap_houses_with_swimming_pool }
+  - match: { hits.hits.1.fields._percolator_document_slot: [ 0 ] }
+  - match: { hits.hits.1.fields._percolator_document_slot_0_matched_queries: [ "swimming_pool_query" ] }

